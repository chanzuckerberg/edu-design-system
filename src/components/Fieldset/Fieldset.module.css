.fieldset__footer {
  display: flex;
  justify-content: space-between;

  margin-top: calc(var(--eds-size-3) / 16 * 1rem);
}

/**
 * The contents of the fieldset. Spaces them apart.
 */
 .fieldset-items > :not(:last-child) {
  margin-bottom: calc(var(--eds-size-2) / 16 * 1rem);
}

/**
 * A label that's rendered as a <legend> for fieldsets.
 * It contains the same characteristics as a label (ability to add flag for optional field, etc),
 * but semantically/stylistically a bit different.
 */
.fieldset-legend {
  margin-bottom: calc(var(--eds-size-3) / 16 * 1rem);
}

/**
 * Label flag to mark whether or not a field is required
 * or optional. Currently a flag is only present for optional fields
 *
 * TODO: remove as deprecated with v16 of EDS
 */
.fieldset-legend__flag {
  font: var(--eds-theme-typography-body-sm);
}

.fieldset-legend__overline {
  display: flex;
  align-items: center;
  gap: calc(var(--eds-size-half) / 16 * 1rem);
}
