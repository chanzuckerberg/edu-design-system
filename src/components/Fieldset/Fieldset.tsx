import clsx from 'clsx';
import type { ComponentPropsWithoutRef, ElementType, ReactNode } from 'react';

import React from 'react';
import styles from './Fieldset.module.css';

type FieldsetProps = {
  /**
   * The contents of the fieldset. We suggest a FieldsetLegend followed by
   * interactive elements.
   *
   * Should be wrapped in a fragment to allow our styling to control the spacing
   * between elements.
   */
  children: ReactNode;
  /**
   * Additional classnames passed in for styling.
   */
  className?: string;
} & React.HTMLAttributes<HTMLFieldSetElement>;

export type FieldsetItemsProps<T extends ElementType> = {
  /**
   * The content of the control elements in the fieldset.
   */
  children: ReactNode;
  /**
   * Type of element the immediate wrapper around the contents should be.
   * @default 'div'
   */
  as?: T;
  /**
   * Additional classnames passed in for styling.
   */
  className?: string;
};

export type FieldsetLegendProps = {
  /**
   * CSS class names that can be appended to the component.
   */
  className?: string;
  /**
   * String to indicate required or optional state.
   */
  optionalLabel?: '(required)' | '(optional)';
  /**
   * Legend text string that names the fieldset.
   */
  text: string;
};

/**
 * `import {Fieldset} from "@chanzuckerberg/eds";`
 *
 * A container for a fieldset that includes a legend and one or more form inputs.
 */
export function Fieldset({ children, className }: FieldsetProps) {
  const componentClassName = clsx(styles['fieldset'], className);
  return <fieldset className={componentClassName}>{children}</fieldset>;
}

/**
 * Helper sub-component for styling the control elements in the component.
 */
export const FieldsetItems = <T extends ElementType = 'div'>({
  children,
  as,
  className,
  ...props
}: FieldsetItemsProps<T> &
  Omit<ComponentPropsWithoutRef<T>, keyof FieldsetItemsProps<T>>) => {
  const componentClassName = clsx(styles['fieldset-items'], className);
  const Component = as || 'div';
  // Disable this once EDS is updated to React 18+ or lib specifies version.
  // There is a type mismatch betwwen what gets pulled in via react-beautiful-dnd
  // and React 16. that library imports the latest react types regardless of
  // installed version.
  return (
    <Component className={componentClassName} {...props}>
      {children}
    </Component>
  );
};

/**
 * Helper sub-component for styling the legend in a fieldset.
 */
const FieldsetLegend = ({
  className,
  optionalLabel,
  text,
  ...other
}: FieldsetLegendProps) => {
  const componentClassName = clsx(styles['fieldset-legend'], className);
  return (
    <legend className={componentClassName} {...other}>
      {text}{' '}
      {optionalLabel && (
        <span className={styles['fieldset-legend__flag']}>{optionalLabel}</span>
      )}
    </legend>
  );
};

Fieldset.Items = FieldsetItems;
Fieldset.Legend = FieldsetLegend;
