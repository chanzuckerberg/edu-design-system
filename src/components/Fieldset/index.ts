export { Fieldset as default } from "./Fieldset";
