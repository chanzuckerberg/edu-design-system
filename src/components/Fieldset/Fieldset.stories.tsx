import type { StoryObj } from '@storybook/react';
import React from 'react';
import { Fieldset } from './Fieldset';
import Checkbox from '../Checkbox';

export default {
  title: 'Atoms/Forms/Fieldset',
  component: Fieldset,
};

type Args = React.ComponentProps<typeof Fieldset>;

export const Default: StoryObj<Args> = {
  args: {
    children: (
      <>
        <Fieldset.Legend>Legend</Fieldset.Legend>
        <Fieldset.Items>
          <Checkbox label="Checkbox label 1" />
          <Checkbox label="Checkbox label 2" />
          <Checkbox label="Checkbox label 3" />
          <Checkbox label="Checkbox label 4" />
          <Checkbox label="Checkbox label 5" />
        </Fieldset.Items>
      </>
    ),
  },
};
