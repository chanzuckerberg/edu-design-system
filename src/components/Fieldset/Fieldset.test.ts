import { generateSnapshots } from '@chanzuckerberg/story-utils';
import type { StoryFile } from '@storybook/testing-react';
import * as stories from './Fieldset.stories';

describe('<Fieldset />', () => {
  generateSnapshots(stories as StoryFile);
});
