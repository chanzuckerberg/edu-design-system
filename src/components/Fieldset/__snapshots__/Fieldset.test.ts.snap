// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fieldset /> Default story renders snapshot 1`] = `
<fieldset
  class="fieldset"
>
  <legend
    class="fieldset-legend"
  >
    Legend
     
  </legend>
  <div
    class="fieldset-items"
  >
    <div
      class="checkbox"
    >
      <span
        class="input__wrapper"
      >
        <input
          class="checkbox__input"
          data-bootstrap-override="checkbox"
          id="1"
          style="--checkbox-svg-size: 1.25rem;"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          height="1.25rem"
          viewBox="0 0 20 20"
          width="1.25rem"
        >
          <rect
            class="checkmark__outline"
            fill="none"
            height="18"
            rx="3"
            stroke="currentColor"
            stroke-width="2"
            width="18"
            x="1"
            y="1"
          />
          <use
            class="checkmark__content"
            fill="none"
            xlink:href="test-file-stub#check"
          />
        </svg>
      </span>
      <label
        class="label label--md"
        data-bootstrap-override="label"
        for="1"
        style="--checkbox-svg-size: 1.25rem;"
      >
        Checkbox label 1
      </label>
    </div>
    <div
      class="checkbox"
    >
      <span
        class="input__wrapper"
      >
        <input
          class="checkbox__input"
          data-bootstrap-override="checkbox"
          id="2"
          style="--checkbox-svg-size: 1.25rem;"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          height="1.25rem"
          viewBox="0 0 20 20"
          width="1.25rem"
        >
          <rect
            class="checkmark__outline"
            fill="none"
            height="18"
            rx="3"
            stroke="currentColor"
            stroke-width="2"
            width="18"
            x="1"
            y="1"
          />
          <use
            class="checkmark__content"
            fill="none"
            xlink:href="test-file-stub#check"
          />
        </svg>
      </span>
      <label
        class="label label--md"
        data-bootstrap-override="label"
        for="2"
        style="--checkbox-svg-size: 1.25rem;"
      >
        Checkbox label 2
      </label>
    </div>
    <div
      class="checkbox"
    >
      <span
        class="input__wrapper"
      >
        <input
          class="checkbox__input"
          data-bootstrap-override="checkbox"
          id="3"
          style="--checkbox-svg-size: 1.25rem;"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          height="1.25rem"
          viewBox="0 0 20 20"
          width="1.25rem"
        >
          <rect
            class="checkmark__outline"
            fill="none"
            height="18"
            rx="3"
            stroke="currentColor"
            stroke-width="2"
            width="18"
            x="1"
            y="1"
          />
          <use
            class="checkmark__content"
            fill="none"
            xlink:href="test-file-stub#check"
          />
        </svg>
      </span>
      <label
        class="label label--md"
        data-bootstrap-override="label"
        for="3"
        style="--checkbox-svg-size: 1.25rem;"
      >
        Checkbox label 3
      </label>
    </div>
    <div
      class="checkbox"
    >
      <span
        class="input__wrapper"
      >
        <input
          class="checkbox__input"
          data-bootstrap-override="checkbox"
          id="4"
          style="--checkbox-svg-size: 1.25rem;"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          height="1.25rem"
          viewBox="0 0 20 20"
          width="1.25rem"
        >
          <rect
            class="checkmark__outline"
            fill="none"
            height="18"
            rx="3"
            stroke="currentColor"
            stroke-width="2"
            width="18"
            x="1"
            y="1"
          />
          <use
            class="checkmark__content"
            fill="none"
            xlink:href="test-file-stub#check"
          />
        </svg>
      </span>
      <label
        class="label label--md"
        data-bootstrap-override="label"
        for="4"
        style="--checkbox-svg-size: 1.25rem;"
      >
        Checkbox label 4
      </label>
    </div>
    <div
      class="checkbox"
    >
      <span
        class="input__wrapper"
      >
        <input
          class="checkbox__input"
          data-bootstrap-override="checkbox"
          id="5"
          style="--checkbox-svg-size: 1.25rem;"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          height="1.25rem"
          viewBox="0 0 20 20"
          width="1.25rem"
        >
          <rect
            class="checkmark__outline"
            fill="none"
            height="18"
            rx="3"
            stroke="currentColor"
            stroke-width="2"
            width="18"
            x="1"
            y="1"
          />
          <use
            class="checkmark__content"
            fill="none"
            xlink:href="test-file-stub#check"
          />
        </svg>
      </span>
      <label
        class="label label--md"
        data-bootstrap-override="label"
        for="5"
        style="--checkbox-svg-size: 1.25rem;"
      >
        Checkbox label 5
      </label>
    </div>
  </div>
</fieldset>
`;

exports[`<Fieldset /> FieldsetLegends story renders snapshot 1`] = `
<legend
  class="fieldset-legend"
>
  Legend
   
</legend>
`;
